internal/imagerewrite/rewrite.go

```diff
@@ -35,45 +35,55 @@ type DockerImageReference struct {
 	Digest      string
 }
 
-// Regex matches prefix@sha256:checksum, prefix:version, and prefix:version@sha256:checksum.
-var ociImagePattern = regexp.MustCompile(
-	`^(?:(?P<host>[\w.-]+(?::\d+)?(?:/[\w.-]+)*)/)?` +
-		`(?P<name>[a-z0-9]+(?:[_-][a-z0-9]+)*)` +
-		`(?:` +
-		`(?::(?P<tag>[\w.\-]+))(?:@(?P<digest>sha256:[a-fA-F0-9]{64}))?|` +
-		`@(?P<digest>sha256:[a-fA-F0-9]{64})` +
-		`)` +
-		`$`,
-)
-
+// NewDockerImageReference parses a Docker image reference string and returns a DockerImageReference struct.
+// It extracts the host, repository path, image name, tag, and digest from the input string.
+// Returns an error if the reference is invalid or cannot be parsed.
 func NewDockerImageReference(val string) (*DockerImageReference, error) {
+	namedRef, err := reference.ParseNormalizedNamed(val)
+	if err != nil {
+		return nil, fmt.Errorf("%w: %s", ErrInvalidImageReference, err.Error())
+	}
+
 	res := &DockerImageReference{}
 
-	matches := ociImagePattern.FindStringSubmatch(val)
-	if matches == nil {
-		return nil, ErrInvalidImageReference
+	domain := reference.Domain(namedRef)
+	path := reference.Path(namedRef)
+
+	familiarName := reference.FamiliarName(namedRef)
+	imageName := familiarName
+	if idx := strings.LastIndex(familiarName, "/"); idx != -1 {
+		imageName = familiarName[idx+1:]
 	}
-	result := map[string]string{}
-	for i, name := range ociImagePattern.SubexpNames() {
-		if matches[i] == "" {
-			// Skip empty captures because there are two "digest" groups and only one will match.
-			continue
-		}
-		result[name] = matches[i]
+
+	// Derive repository path (everything before last segment).
+	repoPath := ""
+	if lastSlash := strings.LastIndex(path, "/"); lastSlash != -1 {
+		repoPath = path[:lastSlash]
 	}
 
-	res.HostAndPath = result["host"]
-	imageName := result["name"]
-	tag := result["tag"]
+	if domain != "" {
+		if repoPath != "" {
+			res.HostAndPath = domain + "/" + repoPath
+		} else {
+			res.HostAndPath = domain
+		}
+	} else {
+		res.HostAndPath = repoPath
+	}
 
+	// Populate NameAndTag and Digest
```

Comment:
If you defined some function with meaningful name, you don't have to leave such comment